test/src/lib/backoff.spec.ts

```diff
@@ -0,0 +1,216 @@
+		const minDelay = Math.floor(Math.random() * 1000);
+		const myBackoffFunc = withBackoff(
+			async () => {
+				return failer.willResolve('ok');
+			},
+			{
+				minDelay,
+			},
+		);
+		// Function should have been called 0 times to start
+		expect(failer.callCount).to.equal(0);
+		// Call function
+		myBackoffFunc();
+		// Check that function is ran atleast once
+		expect(failer.callCount).to.equal(1);
+		// Elapse some time but not enough to be minDelay
+		await clock.tickAsync(minDelay - 1);
+		// Check that the function still has only been called 1 time
+		expect(failer.callCount).to.equal(1);
+		// Elapse exactly minDelay so function is called once more
+		await clock.tickAsync(minDelay + 1);
+		// Check that function was called twice
+		expect(failer.callCount).to.equal(2);
+	});
+
+	it('backs off with exponential delay', async () => {
+		const failer = new Failer(3);
+		const minDelay = Math.floor(Math.random() * 1000);
+		const myBackoffFunc = withBackoff(
+			async () => {
+				return failer.willResolve('ok');
+			},
+			{
+				minDelay,
+				maxDelay: 5000000,
+			},
+		);
+		expect(failer.callCount).to.equal(0);
+		// Call function
+		const p = myBackoffFunc();
+		// Function is called immediately
+		expect(failer.callCount).to.equal(1);
+		// First delay is equal to minDelay
+		await clock.tickAsync(minDelay);
+		// Should have been called again
+		expect(failer.callCount).to.equal(2);
+		// Tick exponential time
+		await clock.tickAsync(minDelay * 2);
+		// Should have been called again
+		expect(failer.callCount).to.equal(3);
+		// Tick exponential time
+		await clock.tickAsync(minDelay * 2 * 2);
+		// Function should be fulfilled by now
+		await expect(p).to.be.fulfilled;
+		expect(failer.callCount).to.equal(4);
+	});
+
+	it('never exceeds maxRetries', async () => {
+		// Make the failer fail 100 times (more then maxRetries)
+		const failer = new Failer(100);
+		const fnWithBackoff = withBackoff(async () => {
+			return failer.willResolve('ok');
+		}, DEFAULT_OPTIONS);
+		clock.runAllAsync();
+		// Call the function
+		await expect(fnWithBackoff()).to.eventually.be.rejectedWith(
+			`Reached max number of retries: ${DEFAULT_OPTIONS.maxRetries}`,
+		);
+	});
+
+	it('provides correct info within onFailure callback', async () => {
+		// Create a function that will fail 3 times so it succeeds on the 4th
+		const failer = new Failer(3);
+		const minDelay = Math.floor(Math.random() * 1000);
+		let counter = 0;
+		// Wrap function withBackoff
+		const fnWithBackoff = withBackoff(
+			async () => {
+				return failer.willResolve('ok');
+			},
+			{
+				minDelay,
+				onFailure: (data: OnFailureInfo) => {
+					counter++;
+					expect(data).to.deep.equal({
+						failures: counter,
+						delay:
+							counter === 1 ? minDelay : exponentialize(minDelay, counter - 1),
+						error: 'Not ready!',
+					});
+				},
+			},
+		);
+		// Call function and allow clock to trigger all events
+		clock.runAllAsync();
```

Comment:
does this work? I have only seen this being done the oppossite way

```
const p = fnWithBackoff();
await clock.runAllAsync();
await p;
```

Reply:
"clock.runAll(); / await clock.runAllAsync()

This runs all pending timers until there are none remaining. If new timers are added while it is executing they will be run as well."